File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-fx-5-4/FxAverageRateObservation.cpp
// FxAverageRateObservation.cpp 
#include "FxAverageRateObservation.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

FxAverageRateObservation::FxAverageRateObservation(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //dateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* dateNode = xmlNode->FirstChildElement("date");

   if(dateNode){dateIsNull_ = false;}
   else{dateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- dateNode , address : " << dateNode << std::endl;
   #endif
   if(dateNode)
   {
      if(dateNode->Attribute("href") || dateNode->Attribute("id"))
      {
          if(dateNode->Attribute("id"))
          {
             dateIDRef_ = dateNode->Attribute("id");
             date_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(dateNode));
             date_->setID(dateIDRef_);
             IDManager::instance().setID(dateIDRef_,date_);
          }
          else if(dateNode->Attribute("href")) { dateIDRef_ = dateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { date_ = boost::shared_ptr<XsdTypeDate>(new XsdTypeDate(dateNode));}
   }

   //averageRateWeightingFactorNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* averageRateWeightingFactorNode = xmlNode->FirstChildElement("averageRateWeightingFactor");

   if(averageRateWeightingFactorNode){averageRateWeightingFactorIsNull_ = false;}
   else{averageRateWeightingFactorIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- averageRateWeightingFactorNode , address : " << averageRateWeightingFactorNode << std::endl;
   #endif
   if(averageRateWeightingFactorNode)
   {
      if(averageRateWeightingFactorNode->Attribute("href") || averageRateWeightingFactorNode->Attribute("id"))
      {
          if(averageRateWeightingFactorNode->Attribute("id"))
          {
             averageRateWeightingFactorIDRef_ = averageRateWeightingFactorNode->Attribute("id");
             averageRateWeightingFactor_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(averageRateWeightingFactorNode));
             averageRateWeightingFactor_->setID(averageRateWeightingFactorIDRef_);
             IDManager::instance().setID(averageRateWeightingFactorIDRef_,averageRateWeightingFactor_);
          }
          else if(averageRateWeightingFactorNode->Attribute("href")) { averageRateWeightingFactorIDRef_ = averageRateWeightingFactorNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { averageRateWeightingFactor_ = boost::shared_ptr<XsdTypeDecimal>(new XsdTypeDecimal(averageRateWeightingFactorNode));}
   }

   //rateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* rateNode = xmlNode->FirstChildElement("rate");

   if(rateNode){rateIsNull_ = false;}
   else{rateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- rateNode , address : " << rateNode << std::endl;
   #endif
   if(rateNode)
   {
      if(rateNode->Attribute("href") || rateNode->Attribute("id"))
      {
          if(rateNode->Attribute("id"))
          {
             rateIDRef_ = rateNode->Attribute("id");
             rate_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(rateNode));
             rate_->setID(rateIDRef_);
             IDManager::instance().setID(rateIDRef_,rate_);
          }
          else if(rateNode->Attribute("href")) { rateIDRef_ = rateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { rate_ = boost::shared_ptr<NonNegativeDecimal>(new NonNegativeDecimal(rateNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeDate> FxAverageRateObservation::getDate()
{
   if(!this->dateIsNull_){
        if(dateIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDate>(IDManager::instance().getID(dateIDRef_));
        }else{
             return this->date_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDate>();
   }
}
boost::shared_ptr<XsdTypeDecimal> FxAverageRateObservation::getAverageRateWeightingFactor()
{
   if(!this->averageRateWeightingFactorIsNull_){
        if(averageRateWeightingFactorIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeDecimal>(IDManager::instance().getID(averageRateWeightingFactorIDRef_));
        }else{
             return this->averageRateWeightingFactor_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeDecimal>();
   }
}
boost::shared_ptr<NonNegativeDecimal> FxAverageRateObservation::getRate()
{
   if(!this->rateIsNull_){
        if(rateIDRef_ != ""){
             return boost::shared_static_cast<NonNegativeDecimal>(IDManager::instance().getID(rateIDRef_));
        }else{
             return this->rate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<NonNegativeDecimal>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-fx-5-4/FxAverageRateObservation.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class FxAverageRateObservation : ISerialized
    {
        public FxAverageRateObservation(XmlNode xmlNode)
        {
            XmlNode dateNode = xmlNode.SelectSingleNode("date");
            
            if (dateNode != null)
            {
                if (dateNode.Attributes["href"] != null || dateNode.Attributes["id"] != null) 
                {
                    if (dateNode.Attributes["id"] != null) 
                    {
                        dateIDRef_ = dateNode.Attributes["id"].Value;
                        XsdTypeDate ob = new XsdTypeDate(dateNode);
                        IDManager.SetID(dateIDRef_, ob);
                    }
                    else if (dateNode.Attributes["href"] != null)
                    {
                        dateIDRef_ = dateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        date_ = new XsdTypeDate(dateNode);
                    }
                }
                else
                {
                    date_ = new XsdTypeDate(dateNode);
                }
            }
            
        
            XmlNode averageRateWeightingFactorNode = xmlNode.SelectSingleNode("averageRateWeightingFactor");
            
            if (averageRateWeightingFactorNode != null)
            {
                if (averageRateWeightingFactorNode.Attributes["href"] != null || averageRateWeightingFactorNode.Attributes["id"] != null) 
                {
                    if (averageRateWeightingFactorNode.Attributes["id"] != null) 
                    {
                        averageRateWeightingFactorIDRef_ = averageRateWeightingFactorNode.Attributes["id"].Value;
                        XsdTypeDecimal ob = new XsdTypeDecimal(averageRateWeightingFactorNode);
                        IDManager.SetID(averageRateWeightingFactorIDRef_, ob);
                    }
                    else if (averageRateWeightingFactorNode.Attributes["href"] != null)
                    {
                        averageRateWeightingFactorIDRef_ = averageRateWeightingFactorNode.Attributes["href"].Value;
                    }
                    else
                    {
                        averageRateWeightingFactor_ = new XsdTypeDecimal(averageRateWeightingFactorNode);
                    }
                }
                else
                {
                    averageRateWeightingFactor_ = new XsdTypeDecimal(averageRateWeightingFactorNode);
                }
            }
            
        
            XmlNode rateNode = xmlNode.SelectSingleNode("rate");
            
            if (rateNode != null)
            {
                if (rateNode.Attributes["href"] != null || rateNode.Attributes["id"] != null) 
                {
                    if (rateNode.Attributes["id"] != null) 
                    {
                        rateIDRef_ = rateNode.Attributes["id"].Value;
                        NonNegativeDecimal ob = new NonNegativeDecimal(rateNode);
                        IDManager.SetID(rateIDRef_, ob);
                    }
                    else if (rateNode.Attributes["href"] != null)
                    {
                        rateIDRef_ = rateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        rate_ = new NonNegativeDecimal(rateNode);
                    }
                }
                else
                {
                    rate_ = new NonNegativeDecimal(rateNode);
                }
            }
            
        
        }
        
    
        #region Date_
        private XsdTypeDate date_;
        public XsdTypeDate Date_
        {
            get
            {
                if (this.date_ != null)
                {
                    return this.date_; 
                }
                else if (this.dateIDRef_ != null)
                {
                    date_ = IDManager.getID(dateIDRef_) as XsdTypeDate;
                    return this.date_; 
                }
                else
                {
                      return this.date_; 
                }
            }
            set
            {
                if (this.date_ != value)
                {
                    this.date_ = value;
                }
            }
        }
        #endregion
        
        public string dateIDRef_ { get; set; }
        #region AverageRateWeightingFactor_
        private XsdTypeDecimal averageRateWeightingFactor_;
        public XsdTypeDecimal AverageRateWeightingFactor_
        {
            get
            {
                if (this.averageRateWeightingFactor_ != null)
                {
                    return this.averageRateWeightingFactor_; 
                }
                else if (this.averageRateWeightingFactorIDRef_ != null)
                {
                    averageRateWeightingFactor_ = IDManager.getID(averageRateWeightingFactorIDRef_) as XsdTypeDecimal;
                    return this.averageRateWeightingFactor_; 
                }
                else
                {
                      return this.averageRateWeightingFactor_; 
                }
            }
            set
            {
                if (this.averageRateWeightingFactor_ != value)
                {
                    this.averageRateWeightingFactor_ = value;
                }
            }
        }
        #endregion
        
        public string averageRateWeightingFactorIDRef_ { get; set; }
        #region Rate_
        private NonNegativeDecimal rate_;
        public NonNegativeDecimal Rate_
        {
            get
            {
                if (this.rate_ != null)
                {
                    return this.rate_; 
                }
                else if (this.rateIDRef_ != null)
                {
                    rate_ = IDManager.getID(rateIDRef_) as NonNegativeDecimal;
                    return this.rate_; 
                }
                else
                {
                      return this.rate_; 
                }
            }
            set
            {
                if (this.rate_ != value)
                {
                    this.rate_ = value;
                }
            }
        }
        #endregion
        
        public string rateIDRef_ { get; set; }
        
    
        
    
    }
    
}

